channels/testing/base.py

```diff
@@ -0,0 +1,31 @@
+from importlib import import_module
+
+from asgiref.testing import ApplicationCommunicator
+from django.conf import settings
+
+from ..auth import login, logout
+from ..db import database_sync_to_async
+
+
+class AuthCommunicator(ApplicationCommunicator):
+    """
+    ApplicationCommunicator subclass with authentication helpers.
+    """
+
+    async def login(self):
+        """
+        Logs a user in if found in the scope.
+        """
+        if self.scope["user"]:
+            # The session might have been set by the SessionMiddleware
+            if not self.scope.get("session"):
+                engine = import_module(settings.SESSION_ENGINE)
+                self.scope["session"] = engine.SessionStore()
```

Comment:
Im not sure we are supposed to modify the scope after it has been used to call the through all the middlewares etc. @andrewgodwin can you clarify on this?

what about instead injecting a session object into the scope (only if a user is passed in the init) when you add the user object in the init?

Reply:
We're working on figuring this out over in #946. Let me work out what we're doing first, then we can actually request changes.

Reply:
OK, we've decided that scopes are immutable from the outside, so this would need to change.